source/beamline/SXRMB/SXRMBBeamline.cpp

```diff
@@ -128,149 +94,261 @@ AMMotorGroupObject *SXRMBBeamline::microprobeSampleStageMotorGroupObject() const
 	return motorGroup_->motorGroupObject("Microprobe Stage - X, Z, Y");
 }
 
-AMReadOnlyPVControl* SXRMBBeamline::beamlineStatus() const {
+AMReadOnlyPVControl* SXRMBBeamline::beamlineStatus() const
+{
 	return beamlineStatus_;
 }
 
-bool SXRMBBeamline::isConnected() const{
-	return energy_->isConnected() && beamlineStatus_->isConnected() && microprobeSampleStageControlSet_->isConnected();
+bool SXRMBBeamline::isConnected() const
+{
+	return energy_->isConnected() && beamlineStatus_->isConnected()
+			&& microprobeSampleStageControlSet_->isConnected() && beamlineControlShutterSet_->isConnected();
 }
 
-CLSBasicScalerChannelDetector* SXRMBBeamline::i0Detector() const{
+CLSBasicScalerChannelDetector* SXRMBBeamline::i0Detector() const
+{
 	return i0Detector_;
 }
 
-CLSBasicScalerChannelDetector* SXRMBBeamline::teyDetector() const{
+CLSBasicScalerChannelDetector* SXRMBBeamline::teyDetector() const
+{
 	return teyDetector_;
 }
 
-AMBasicControlDetectorEmulator* SXRMBBeamline::energyFeedbackDetector() const{
+AMBasicControlDetectorEmulator* SXRMBBeamline::energyFeedbackDetector() const
+{
 	return energyFeedbackDetector_;
 }
 
-SXRMBBrukerDetector* SXRMBBeamline::brukerDetector() const{
+SXRMBBrukerDetector* SXRMBBeamline::brukerDetector() const
+{
 	return brukerDetector_;
 }
 
-AMAction3* SXRMBBeamline::createBeamOnActions() const{
+AMAction3* SXRMBBeamline::createBeamOnActions() const
+{
 	if(!isConnected())
 		return 0;
 
-	AMListAction3 *beamOnActionsList = new AMListAction3(new AMListActionInfo3("SXRMB Beam On", "SXRMB Beam On"), AMListAction3::Parallel);
+	// if all the valves are already open, we don't need to do that again
+	if (VVR16064B1003Valve_->isOpen() && VVR16064B1004Valve_->isOpen() && VVR16064B1006Valve_->isOpen() && VVR16064B1007Valve_->isOpen() && VVR16065B1001Valve_->isOpen() && PSH1406B1002Shutter_->isOpen())
+		return 0;
+
+	// stage 1: open / wait the valves action list
+	AMListAction3 *valveOpenActionsList = new AMListAction3(new AMListActionInfo3("SXRMB Valve Open action list", "SXRMB Valve Open"), AMListAction3::Sequential);
+	AMListAction3 *valveWaitActionsList = new AMListAction3(new AMListActionInfo3("SXRMB Valve Wait action list", "SXRMB Valve Wait"), AMListAction3::Parallel);
+
+	if (VVR16064B1003Valve_->isClosed()) {
+		AMControlInfo VVR16064B1003ValveSetpoint = VVR16064B1003Valve_->toInfo();
+		VVR16064B1003ValveSetpoint.setValue(1);
+		AMControlMoveAction3 *VVR16064B1003ValveSetpointOpenAction = new AMControlMoveAction3(new AMControlMoveActionInfo3(VVR16064B1003ValveSetpoint), VVR16064B1003Valve_);
```

Comment:
@XuetaoLiu-CLS 
You can probably use the AMActionSupport::buildControlMoveAction() calls for all of these. It's a much cleaner and shorter way to do it.